core/src/main/java/org/apache/calcite/runtime/SpatialTypeTableFunctions.java

```diff
@@ -0,0 +1,158 @@
+import org.apache.calcite.linq4j.tree.Types;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.schema.QueryableTable;
+import org.apache.calcite.schema.ScannableTable;
+import org.apache.calcite.schema.Schema;
+import org.apache.calcite.schema.Schema.TableType;
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.Statistic;
+import org.apache.calcite.schema.Statistics;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.locationtech.jts.geom.Geometry;
+
+import java.lang.reflect.Method;
+
+/**
+ * Helper methods to implement spatial type (ST) table functions.
+ */
+@Deterministic
+@Strict
+@Experimental
+public class SpatialTypeTableFunctions {
+
+  private SpatialTypeTableFunctions() {}
+
+  public static final Method EXPLODE =
+      Types.lookupMethod(SpatialTypeTableFunctions.class,
+          "explode", Geometry.class);
+
+  public static final Method EXPLODE_CURSOR =
+      Types.lookupMethod(SpatialTypeTableFunctions.class,
+          "explodeCursor", Enumerable.class);
+
+  public static ScannableTable explode(Geometry geometry) {
+    return new ScannableTable() {
+
+      @Override public RelDataType getRowType(RelDataTypeFactory typeFactory) {
+        return typeFactory.builder()
+            .add("geom", SqlTypeName.GEOMETRY)
+            .build();
+      }
+
+      @Override public Enumerable<@Nullable Object[]> scan(DataContext root) {
+        return new GeometryElements(geometry).select(g -> new Object[] {g});
+      }
+
+      @Override public Statistic getStatistic() {
+        return Statistics.UNKNOWN;
+      }
+
+      @Override public TableType getJdbcTableType() {
+        return Schema.TableType.TABLE;
+      }
+
+      @Override public boolean isRolledUp(String column) {
+        return false;
+      }
+
+      @Override public boolean rolledUpColumnValidInsideAgg(String column, SqlCall call,
+          @Nullable SqlNode parent, @Nullable CalciteConnectionConfig config) {
+        return true;
+      }
+    };
+  }
+
+  public static QueryableTable explodeCursor(final Enumerable<Object[]> a) {
+    return new AbstractQueryableTable(Object[].class) {
+      @Override public RelDataType getRowType(RelDataTypeFactory typeFactory) {
+        return typeFactory.builder()
+            .add("geom", SqlTypeName.GEOMETRY)
+            .build();
+      }
+
+      @Override public <T> Queryable<T> asQueryable(QueryProvider queryProvider,
+          SchemaPlus schema, String tableName) {
+        final Enumerable<Geometry> enumerable = a.selectMany(a0 -> {
+          final Geometry geometry = (Geometry) a0[0];
+          return new GeometryElements(geometry);
+        });
+        return (Queryable) enumerable.asQueryable();
+      }
+    };
+  }
+
+  /**
```

Comment:
How about using the following?
```
Linq4j.asEnumerable(
  Functions.generate(geometry.getNumGeometries(),
      i -> geometry.getGeometryN(i)));
```

It seems to capture the random-access-generating-list nature, and yours is rather more specific to Geometry than it needs to be. Also it has optimized implementations of `skip` , `take`, `elementAt`, `count`, etc.

Reply:
Thank you for the pointer on Linq4j, this has been fixed.